tree/ntuple/v7/src/RNTupleModel.cxx

```diff
@@ -22,8 +22,109 @@
 #include <atomic>
 #include <cstdlib>
 #include <memory>
+#include <typeinfo>
 #include <utility>
 
+ROOT::Experimental::RResult<void>
+ROOT::Experimental::RNTupleModel::RProjectedFields::EnsureValidMapping(const Detail::RFieldBase *target,
+                                                                       const FieldMap_t &fieldMap)
+{
+   auto source = fieldMap.at(target);
+   if (source->GetStructure() != target->GetStructure())
+      return R__FAIL("field mapping structural mismatch: " + source->GetName() + " --> " + target->GetName());
+   if (target->GetStructure() == ENTupleStructure::kLeaf) {
+      if (typeid(*target) != typeid(*source))
+         return R__FAIL("field mapping type mismatch: " + source->GetName() + " --> " + target->GetName());
+   }
+
+   // We support projections only across records and collections. In the following, we check that the projected
+   // field is on the same path of collection fields in the field tree than the source field.
+
+   // Finds the first non-record parent field of the input field
+   auto fnBreakPoint = [](const Detail::RFieldBase *f) -> const Detail::RFieldBase * {
+      auto parent = f->GetParent();
+      while (parent) {
+         if (parent->GetStructure() != ENTupleStructure::kRecord)
+            return parent;
+         parent = parent->GetParent();
+      }
+      // We reached the zero field
+      return nullptr;
+   };
+
+   // If source or target has a variant or reference as a parent, error out
+   auto *sourceBreakPoint = fnBreakPoint(source);
+   if (sourceBreakPoint && sourceBreakPoint->GetStructure() != ENTupleStructure::kCollection)
+      return R__FAIL("unsupported field mapping (source structure)");
+   auto *targetBreakPoint = fnBreakPoint(target);
+   if (targetBreakPoint && sourceBreakPoint->GetStructure() != ENTupleStructure::kCollection)
+      return R__FAIL("unsupported field mapping (target structure)");
+
+   if (!sourceBreakPoint && !targetBreakPoint) {
+      // Source and target have no collections as parent
+      return RResult<void>::Success();
+   }
+   if (sourceBreakPoint && targetBreakPoint) {
+      if (sourceBreakPoint == targetBreakPoint) {
+         // Source and target are children of the same collection
+         return RResult<void>::Success();
+      }
+      if ((fieldMap.count(targetBreakPoint) > 0) && fieldMap.at(targetBreakPoint) == sourceBreakPoint) {
```

Comment:
Perhaps use `std::unordered_map::find()` to avoid two lookups.

Reply:
[done](https://github.com/root-project/root/pull/11973/commits/f519b9dc677ec89ad62385b8ed3ae74f4f889837)